hal/reset_reason_api.h

```diff
@@ -0,0 +1,67 @@
+/** \addtogroup hal */
+/** @{*/
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_RESET_REASON_API_H
+#define MBED_RESET_REASON_API_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef enum {
+  RESET_REASON_POWER_ON  = (1 << 0), /**< Set when power is initially applied to the board. The power-on-reset circuit causes a POWER_ON reset when this occurs */
+  RESET_REASON_PIN_RESET = (1 << 1), /**< Set when a reset is triggered by the hardware pin on the board */
+  RESET_REASON_BROWN_OUT = (1 << 2), /**< Triggered when the voltage drops below the low voltage detect (LVD) threshold the system will be held in a reset until the voltage rises above the threshold */
+  RESET_REASON_SOFTWARE  = (1 << 3), /**< Set during software reset, typically triggered by writing the SYSRESETREQ bit in the Application Interrupt and Reset Control register */
+  RESET_REASON_WATCHDOG  = (1 << 4), /**< Set when a running watchdog timer fails to be refreshed */
+  RESET_REASON_LOCKUP    = (1 << 5), /**< Set when the core is locked because of an unrecoverable exception */
+  RESET_REASON_MULTIPLE  = (1 << 6), /**< Set if multiple reset reasons are set within the board. Occurs when the reset reason registers aren't cleared between resets */
+  RESET_REASON_PLATFORM  = (1 << 7), /**< Platform specific reset reason not captured in this enum */
+  RESET_REASON_UNKNOWN   = (1 << 8)  /**< Unknown or unreadable reset reason **/
+} reset_reason_t;
+
+/**
+ * Fetch the reset reason for the last system reset
+ *
+ * Note: Some platforms contain reset reason registers that persist through
+ * system resets. If the registers haven't been cleared before calling this
+ * function multiple reasons may be set within the registers. If multiple reset
+ * reasons are detected this function will return RESET_REASON_MULTIPLE.
+ *
+ * @return enum containing the last reset reason for the board.
```

Comment:
Is that accurate? We can have reasons from previous resets as well. I think having separate get/clear calls isn't necessary and will lead to issues. Maybe we should clear the reason automatically here? And only return the reset reason on the first call? @pan- @0xc0170 @scartmell-arm what do you think?

Reply:
It depends on the platform. The K64F boards only provide a function `RCM_GetPreviousResetSources` which only returns the previous reset and doesn't persist through reboots. Which means clearing the flag is effectively useless/impossible. There is a feature for persistent flags that can be cleared called `RCM_GetPreviousResetSources` but it isn't supported on the boards I tested. The other platforms do persist reset sources through reset.

I would advocate clearing the flags implicitly after the function has been read if supported by the platform. But, would say it might be useful to support returning the reason through multiple calls by storing them locally on the first call.

Reply:
I wouldn't assume K64F to be a representative board of the market. At the same time I don't have strong opinion about the clear being explicit/implicit.

Reply:
The reason may also be cleared automatically by the peripheral after it has been read. From my perspective it would be fine if the first call to ` hal_reset_reason_get` since reset is the only call returning a valid value; depending on the architecture, the value may or may not persist; we don't care.

State bookkeeping and clearing logic may be built on top of the get and clear primitives in a platform independant fashion.

Reply:
> State bookkeeping and clearing logic may be built on top of the get and clear primitives in a platform independant fashion.

@scartmell-arm How would this be used from a user space? I do not fully understand the note above, how this get/clear is generic, and I can built on top of it state bookkeeping as noted in the quote
2. clear

Reply:
One of the issues is that the reset reason behaviour is not consistent across the platforms, some platforms do not persist the reset state between resets, and some may implicitly delete the state after it is first read.

It would make sense to implicitly clear the reset registers after they are read to keep everything consistent across the platforms and not require a separate call to clear.

But, the state of the registers should be tracked/stored somewhere to allow multiple reads. e.g if get clears the registers but returns `RESET_REASON_PLATFORM` there is no way to read the registers again to determine the reason without storing the register state.

Reply:
Again that makes sense :) If reading the reset reason clears it on some platforms, we should make it an default behavior and store the reason in a static variable in the Mbed OS driver (rather than HAL).

Reply:
If the reset flags are cleared on first read, then applications with a bootloader, like cloud client, may never know the reset reason if the bootloader uses the watchdog.

Reply:
There's not much that can be done in that case for platforms that clear on the first read, the reset-reason API is currently only configured to clear them through the driver API to ensure they don't persist through reboots for an application that is actually using them.